ApiBundle/Transformer/AreaTransformer.php

```diff
@@ -220,39 +257,81 @@ public function transformFromTemplate(AreaInterface $area, TemplateInterface $te
      * @param AreaInterface|null $source
      * @param NodeInterface|null $node
      *
-     * @return mixed|AreaInterface
+     * @return AreaInterface
      *
      * @throws UnexpectedValueException
      */
     public function reverseTransform(FacadeInterface $facade, $source = null, NodeInterface $node = null)
     {
-        $blocks = $facade->getBlocks();
-        $blockDocument = array();
-
         if (!$source instanceof AreaInterface) {
-            throw new UnexpectedValueException("Area Transformer must be an instance of AreaInterface");
+            throw new UnexpectedValueException("source must be an instance of AreaInterface");
         }
-
-        foreach ($blocks as $position => $blockFacade) {
-            $blockArray = $this->getTransformer('block')->reverseTransformToArray($blockFacade, $node);
-            $blockDocument[$position] = $blockArray;
-            $block = $node->getBlock($blockArray['blockId']);
-            if ($blockArray['nodeId'] !== 0) {
-                $siteId = $this->currentSiteManager->getCurrentSiteId();
-                $blockNode = $this->nodeRepository
-                    ->findInLastVersion($blockArray['nodeId'], $node->getLanguage(), $siteId);
-                $block = $blockNode->getBlock($blockArray['blockId']);
+        $subAreaFacade = $facade->getAreas();
+        $newOrderSubAreas = array();
+        /** @var AreaInterface $subArea */
+        foreach ($source->getAreas() as $subArea) {
+            $order = $this->getAreaOrderInChildren($subArea->getAreaId(), $subAreaFacade);
+            $newOrderSubAreas[$order] = $subArea;
+        }
+        ksort($newOrderSubAreas);
+        $source->setAreas(new ArrayCollection($newOrderSubAreas));
+
+        if ($node instanceof NodeInterface) {
+            $blocks = $facade->getBlocks();
+            $blockDocument = array();
+            foreach ($blocks as $position => $blockFacade) {
+                $blockArray = $this->getTransformer('block')->reverseTransformToArray($blockFacade, $node);
```

Comment:
@alavieille : can you use a dedicated method for each iteration, it would help to quickly understand what is done in the loop